vfs/httpfs/union/union.go

```diff
@@ -0,0 +1,128 @@
+// Package union offers a simple http.FileSystem that can unify multiple filesystems at various mount points.
+package union
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/shurcooL/go/vfs/httpfs/vfsutil"
+)
+
+// New creates an union filesystem with the provided mapping of mount points to filesystems.
+//
+// Each mount point must be of form "/mydir". It must start with a '/', and contain a single directory name.
+func New(mapping map[string]http.FileSystem) http.FileSystem {
+	u := &unionFS{
+		ns: make(map[string]http.FileSystem),
+		root: &dirInfo{
+			name:    "/",
+			entries: []os.FileInfo{}, // Not nil.
+			modTime: time.Time{},     // An empty union filesystem has never been modified.
+		},
+	}
+	for mountPoint, fs := range mapping {
+		u.bind(mountPoint, fs)
+	}
+	return u
+}
+
+type unionFS struct {
+	ns   map[string]http.FileSystem // Key is mount point, e.g., "/mydir".
+	root *dirInfo
+}
+
+// bind mounts fs at mountPoint.
+// mountPoint must be of form "/mydir". It must start with a '/', and contain a single directory name.
+func (u *unionFS) bind(mountPoint string, fs http.FileSystem) {
+	fi, err := vfsutil.Stat(fs, "/")
+	if err != nil {
+		log.Fatalln("can't stat root directory of provided filesystem:", err)
+	}
+	modTime := fi.ModTime()
+
+	u.ns[mountPoint] = fs
```

Comment:
No, it couldn't happen. `bind` is unexported, and only called when `unionFS` is created in `New`.

Otherwise, this would need synchronization.